test/hotspot/jtreg/compiler/c2/irTests/TestFloat16ScalarOperations.java

```diff
@@ -0,0 +1,494 @@
+        Float16 actual = valueOf(0.0f);
+        for (int i = 0; i < 50; i++) {
+            Float16 divisor = valueOf(8.0f);
+            Float16 dividend = shortBitsToFloat16(src[i]);
+            actual = add(actual, divide(dividend, divisor));
+            divisor = valueOf(16.0f);
+            actual = add(actual, divide(dividend, divisor));
+            divisor = valueOf(32.0f);
+            actual = add(actual, divide(dividend, divisor));
+        }
+        float expected = 0.0f;
+        for (int i = 0; i < 50; i++) {
+            float dividend = Float.float16ToFloat(src[i]);
+            expected += dividend / 8.0f;
+            expected += dividend / 16.0f;
+            expected += dividend / 32.0f;
+        }
+        assertResult(Math.round(actual.floatValue()), Math.round(expected), "testDivByPOT");
+    }
+
+    @Test
+    @IR(counts = {IRNode.MUL_HF, " 0 ", IRNode.ADD_HF, " >0 ", IRNode.REINTERPRET_S2HF, " >0 ", IRNode.REINTERPRET_HF2S, " >0 "},
+        applyIfCPUFeature = {"avx512_fp16", "true"})
+    public void testMulByTWO() {
+        Float16 res = valueOf(0.0f);
+        Float16 multiplier = valueOf(2.0f);
+        for (int i = 0; i < 20; i++) {
+            Float16 multiplicand = valueOf((float)i);
+            res = add(res, multiply(multiplicand, multiplier));
+        }
+        assertResult(res.floatValue(), (float)((20 * (20 - 1))/2) * 2.0f, "testMulByTWO");
+    }
+
+
+    //
+    // Tests points for various Float16 constant folding transforms. Following figure represents various
+    // special IEEE 754 binary16 values on a number line
+    //
+    //   -Inf                               -0.0                              Inf
+    //   -------|-----------------------------|----------------------------|------
+    //     -MAX_VALUE                        0.0                        MAX_VALUE
+    //
+    //  Number whose exponent lie between -14 and 15, both values inclusive, belongs to normal value range.
+    //  IEEE 754 binary16 specification allows graceful degradation of numbers with exponents less than -14
+    //  into a sub-normal value range i.e. their exponents may extend uptill -24, this is because format
+    //  supports 10 mantissa bits which can be used to represent a number with exponents less than -14.
+    //
+    //  A number below the sub-normal value range is considered as 0.0. With regards to overflowing
+    //  semantics, a value equal to or greater than MAX_VALUE + half ulp (MAX_VALUE) is considered as
+    //  an Infinite value on both side of axis.
+    //
+    //  In addition, format specifies special bit representation for +Inf, -Inf and NaN values.
+    //
+    //  Tests also covers special cases for various operations as per Java SE specification.
+    //
+
+
+    @Test
+    @IR(counts = {IRNode.ADD_HF, " 0 ", IRNode.REINTERPRET_S2HF, " 0 ", IRNode.REINTERPRET_HF2S, " 0 "},
+        applyIfCPUFeature = {"avx512_fp16", "true"})
+    public void testAddConstantFolding() {
```

Comment:
Ok, this is great. I'm missing some cases that check correct rounding. For that, it might be good to have one example with random constants, so 2 random Float16 values. You can generate them in static context, and also compute the result in static context, so it should be evaluated in the interpreter. That way, we can compare the result of interpreter to compiled code.

Do that for all operations.

Reply:
Hey @eme64 , constant folding is done at FP32 granularity, so we first upcast FP16 to FP32 values using hf2f runtime helper, operate over FP32 values and then down cast it back to FP16 value using f2hf helper.  Thus both compiler value transformations and interpreter use the same runtime helper underneath.

Fallback implementation of each Float16 API is using Float.floatToFloat16 and Float.floa16ToFloat routines which are intrinsified at [interpreter level.](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/interpreter/templateInterpreterGenerator.cpp#L488), these interpreter intrinsic invokes same leaf level macro assembly routine flt16_to_flt which is also called though runtime helpers. 

So it may not add much value to do interpreter vs compiler comparison in these cases.

Reply:
Ah, yes, you are right, compiler vs interpreter comparison does not help as much as I thought, though we should still do it. What we need to do is compare interpreter and C2-constant-folded results with the results of the backend instructions, but we can also do that with variable values.